pkg/database/monitors.go

```diff
@@ -0,0 +1,179 @@
+}
+
+// NewMonitors returns a new Monitors
+func NewMonitors(ctx context.Context, uuid string, dbc cosmosdb.DatabaseClient, dbid, collid string) (Monitors, error) {
+	collc := cosmosdb.NewCollectionClient(dbc, dbid)
+
+	triggers := []*cosmosdb.Trigger{
+		{
+			ID:               "renewLease",
+			TriggerOperation: cosmosdb.TriggerOperationAll,
+			TriggerType:      cosmosdb.TriggerTypePre,
+			Body: `function trigger() {
+	var request = getContext().getRequest();
+	var body = request.getBody();
+	var date = new Date();
+	body["leaseExpires"] = Math.floor(date.getTime() / 1000) + 60;
+	request.setBody(body);
+}`,
+		},
+	}
+
+	triggerc := cosmosdb.NewTriggerClient(collc, collid)
+	for _, trigger := range triggers {
+		_, err := triggerc.Create(ctx, trigger)
+		if err != nil && !cosmosdb.IsErrorStatusCode(err, http.StatusConflict) {
+			return nil, err
+		}
+	}
+
+	return &monitors{
+		c:    cosmosdb.NewMonitorDocumentClient(collc, collid),
+		uuid: uuid,
+	}, nil
+}
+
+func (c *monitors) Create(ctx context.Context, doc *api.MonitorDocument) (*api.MonitorDocument, error) {
+	if doc.ID != strings.ToLower(doc.ID) {
+		return nil, fmt.Errorf("id %q is not lower case", doc.ID)
+	}
+
+	doc, err := c.c.Create(ctx, doc.ID, doc, nil)
+
+	if err, ok := err.(*cosmosdb.Error); ok && err.StatusCode == http.StatusConflict {
+		err.StatusCode = http.StatusPreconditionFailed
+	}
+
+	return doc, err
+}
+
+func (c *monitors) get(ctx context.Context, id string) (*api.MonitorDocument, error) {
+	if id != strings.ToLower(id) {
+		return nil, fmt.Errorf("id %q is not lower case", id)
+	}
+
+	return c.c.Get(ctx, id, id, nil)
+}
+
+func (c *monitors) patch(ctx context.Context, id string, f func(*api.MonitorDocument) error, options *cosmosdb.Options) (*api.MonitorDocument, error) {
+	var doc *api.MonitorDocument
+
+	err := cosmosdb.RetryOnPreconditionFailed(func() (err error) {
+		doc, err = c.get(ctx, id)
+		if err != nil {
+			return
+		}
+
+		err = f(doc)
+		if err != nil {
+			return
+		}
+
+		doc, err = c.update(ctx, doc, options)
+		return
+	})
+
+	return doc, err
+}
+
+func (c *monitors) PatchWithLease(ctx context.Context, id string, f func(*api.MonitorDocument) error) (*api.MonitorDocument, error) {
+	return c.patch(ctx, id, func(doc *api.MonitorDocument) error {
+		if doc.LeaseOwner != c.uuid {
+			return fmt.Errorf("lost lease")
+		}
+
+		return f(doc)
+	}, &cosmosdb.Options{PreTriggers: []string{"renewLease"}})
+}
+
+func (c *monitors) update(ctx context.Context, doc *api.MonitorDocument, options *cosmosdb.Options) (*api.MonitorDocument, error) {
+	if doc.ID != strings.ToLower(doc.ID) {
+		return nil, fmt.Errorf("id %q is not lower case", doc.ID)
+	}
+
+	return c.c.Replace(ctx, doc.ID, doc, options)
+}
+
+func (c *monitors) Dequeue(ctx context.Context) (*api.MonitorDocument, error) {
+	docs, err := c.c.QueryAll(ctx, "", &cosmosdb.Query{
+		Query: `SELECT * FROM Monitors doc WHERE doc.id = "monitor" AND (doc.leaseExpires ?? 0) < GetCurrentTimestamp() / 1000`,
```

Comment:
`doc.id = "monitor"` is confusing. Can we document logic little bit?  Considering below we have `SELECT * FROM Monitors doc WHERE doc.id != "monitor"`

Reply:
Yeah, the Dequeue name really doesn't help as well.  I'll have another go at the terminology and comments.